File: internal/scan/ignore.go
package scan

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/umeraamir69/secsentry/internal/gitutil"
)

const (
	ignoreFile    = ".secsentryignore"
	allowlistFile = ".secsentryallow"
)

var skipDirs = map[string]bool{
	".git": true, "node_modules": true, ".venv": true, "venv": true,
	"__pycache__": true, "dist": true, "build": true,
	".mypy_cache": true, ".pytest_cache": true, ".ruff_cache": true,
	".secsentry": true,
}

type ignores struct {
	patterns []string
}

func loadPatterns(path string) []string {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	var out []string
	for _, line := range strings.Split(string(raw), "\n") {
		line = strings.TrimSpace(line)
		if line != "" && !strings.HasPrefix(line, "#") {
			out = append(out, line)
		}
	}
	return out
}

func loadIgnores(repo string) ignores {
	return ignores{patterns: loadPatterns(filepath.Join(repo, ignoreFile))}
}

func (ig ignores) match(rel string) bool {
	rel = strings.ReplaceAll(rel, "\\", "/")
	base := filepath.Base(rel)
	for _, pat := range ig.patterns {
		if ok, _ := filepath.Match(pat, rel); ok {
			return true
		}
		if ok, _ := filepath.Match(pat, base); ok {
			return true
		}
		if strings.HasSuffix(pat, "/") && strings.HasPrefix(rel, pat) {
			return true
		}
		if strings.Contains(rel, pat) && strings.ContainsAny(pat, "*?[") {
			continue
		}
	}
	return false
}

func loadAllowlist(repo string) []string {
	var out []string
	for _, p := range loadPatterns(filepath.Join(repo, allowlistFile)) {
		fp := strings.Fields(p)[0]
		if fp != "" {
			out = append(out, strings.ToLower(fp))
		}
	}
	return out
}

func isAllowlisted(fingerprint string, allowed []string) bool {
	fp := strings.ToLower(fingerprint)
	for _, a := range allowed {
		if a != "" && strings.HasPrefix(fp, a) {
			return true
		}
	}
	return false
}

func skippedPath(rel string) bool {
	for _, part := range strings.Split(strings.ReplaceAll(rel, "\\", "/"), "/") {
		if skipDirs[part] {
			return true
		}
	}
	return false
}

func candidatePaths(root string) []string {
	if gitutil.IsRepo(root) {
		if tracked, err := gitutil.TrackedFiles(root); err == nil && tracked != nil {
			return tracked
		}
	}
	var out []string
	_ = filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		rel, _ := filepath.Rel(root, path)
		if d.IsDir() {
			if skipDirs[d.Name()] {
				return filepath.SkipDir
			}
			return nil
		}
		if skippedPath(rel) {
			return nil
		}
		out = append(out, filepath.ToSlash(rel))
		return nil
	})
	return out
}

File: src/secsentry/scan/ignore.py
"""Which files to skip, and which findings the user has accepted.

Inside a git repo we ask git itself which files are ignored, so `.gitignore`
is honoured exactly rather than reimplemented. `.secsentryignore` adds
scanner-only patterns on top, and the allowlist suppresses findings by
fingerprint — never by storing the secret.
"""

from __future__ import annotations

import fnmatch
from pathlib import Path

from secsentry.git.run import git

SKIP_DIRS = {
    ".git",
    "node_modules",
    ".venv",
    "venv",
    "__pycache__",
    "dist",
    "build",
    ".mypy_cache",
    ".pytest_cache",
    ".ruff_cache",
    ".secsentry",
}

IGNORE_FILE = ".secsentryignore"
ALLOWLIST_FILE = ".secsentryallow"


def _read_patterns(path: Path) -> list[str]:
    if not path.is_file():
        return []
    out = []
    for raw in path.read_text(encoding="utf-8", errors="replace").splitlines():
        line = raw.strip()
        if line and not line.startswith("#"):
            out.append(line)
    return out


class Ignores:
    def __init__(self, repo: Path) -> None:
        self.patterns = _read_patterns(repo / IGNORE_FILE)

    def match(self, rel: str) -> bool:
        rel = rel.replace("\\", "/")
        for pat in self.patterns:
            if fnmatch.fnmatch(rel, pat) or fnmatch.fnmatch(Path(rel).name, pat):
                return True
            if pat.endswith("/") and rel.startswith(pat):
                return True
        return False


def load_allowlist(repo: Path) -> set[str]:
    """Accepted findings, identified by SHA-256 fingerprint prefix."""
    return {p.split()[0].lower() for p in _read_patterns(repo / ALLOWLIST_FILE)}


def is_allowlisted(fingerprint: str, allowed: set[str]) -> bool:
    return any(fingerprint.startswith(a) for a in allowed if a)


def git_tracked_files(repo: Path) -> list[str] | None:
    """Files git would show: tracked plus untracked, minus everything ignored."""
    try:
        out = git(repo, "ls-files", "--cached", "--others", "--exclude-standard")
    except Exception:
        return None
    return [line for line in out.splitlines() if line.strip()]
